zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java

```diff
@@ -0,0 +1,373 @@
+   * @param restoreTempSnapLog A FileTxnSnapLog instance created on the specified local temporary directory path
+   * @throws IOException
+   */
+  private void copyRestoredFilesToLocalTempDir(FileTxnSnapLog restoreTempSnapLog)
+      throws IOException {
+    // Copy backup files to local temp directory
+    for (BackupFileInfo backedupFile : filesToCopy) {
+      String standardFilename = backedupFile.getStandardFile().getName();
+      // Does not matter if it's dataDir or logDir since we use same path for these two directories
+      File localTempDest = new File(restoreTempSnapLog.getDataDir(), standardFilename);
+
+      if (!localTempDest.exists()) {
+        LOG.info("Copying " + backedupFile.getBackedUpFile() + " from backup storage to temp dir "
+            + localTempDest.getPath() + ".");
+        storage.copyToLocalStorage(backedupFile.getBackedUpFile(), localTempDest);
+      } else {
+        LOG.info(
+            "Skipping copying " + backedupFile.getBackedUpFile() + " because it already exists as "
+                + localTempDest.getPath() + ".");
+      }
+    }
+  }
+
+  /**
+   * Process the restored files stored in local restore temporary directory to get them ready to be copied to final destination
+   * The processing currently includes truncating txn logs to zxidToRestore.
+   * @param restoreTempSnapLog A FileTxnSnapLog instance created on the specified local temporary directory path
+   * @param zxidToRestore The zxid to restore to
+   * @throws IOException
+   */
+  private void processRestoredFiles(FileTxnSnapLog restoreTempSnapLog, long zxidToRestore) {
+    if (zxidToRestore != Long.MAX_VALUE) {
+      restoreTempSnapLog.truncateLog(zxidToRestore);
+      LOG.info(
+          "Successfully truncate the logs inside restoreTempDir " + restoreTempDir + " to zxid "
+              + zxidToRestore);
+    }
+  }
+
+  /**
+   * Copy the processed files from local restore temp directory to final destination:
+   * snapshot to snapDir, txn logs to logDir.
+   * @param restoreTempSnapLog A FileTxnSnapLog instance created on the specified local temporary directory path
+   * @throws IOException
+   */
+  private void copyProcessedRestoredFilesToDestination(FileTxnSnapLog restoreTempSnapLog)
+      throws IOException {
+    Stream<Path> processedFilePaths = Files.list(restoreTempSnapLog.getDataDir().toPath());
+    List<File> processedFiles =
+        processedFilePaths.map(filePath -> new File(String.valueOf(filePath)))
+            .collect(Collectors.toList());
+    for (File processedFile : processedFiles) {
+      String fileName = processedFile.getName();
+      File finalDestinationBase = null;
+      if (Util.isSnapshotFileName(fileName)) {
+        finalDestinationBase = snapLog.getSnapDir();
+      } else if (Util.isLogFileName(fileName)) {
+        finalDestinationBase = snapLog.getDataDir();
+      }
+      if (finalDestinationBase != null) {
```

Comment:
If `finalDestinationBase` is `null`, do we just do nothing? Or can it ever be null? I might be misunderstanding it, but I feel like we could tighten the logic here a little more.

Reply:
It does not mean something went wrong - the only case I can think of (but actually exists) is that "initialize" file being created by `ClientBase.createTmpDir();` that we used in the tests, and that's why I just skip this file and do nothing here. Let me sync up offline with you and let's brainstorm if there are other cases for this situation.

Reply:
In that case,  you could do the check for isSnapshot or isLog, if neither, you can just simply `continue` :)